python/ray/tune/sync_client.py

```diff
@@ -0,0 +1,141 @@
+        self.sync_down_func(source, target)
+        return True
+
+
+def no_op(source, target):
+    return
+
+
+NOOP = FunctionBasedClient(no_op, no_op)
+
+
+class CommandBasedClient(SyncClient):
+    def __init__(self, sync_up_template, sync_down_template):
+        """Syncs between two directories with the given command.
+
+        Arguments:
+            sync_up_template (str): A runnable string template; needs to
+                include replacement fields '{source}' and '{target}'.
+            sync_down_template (str): A runnable string template; needs to
+                include replacement fields '{source}' and '{target}'.
+        """
+        self._validate_sync_string(sync_up_template)
+        self._validate_sync_string(sync_down_template)
+        self.sync_up_template = sync_up_template
+        self.sync_down_template = sync_down_template
+        self.logfile = None
+        self.sync_process = None
+
+    def set_logdir(self, logdir):
+        """Sets the directory to log sync execution output in.
+
+        Args:
+            logdir (str): Log directory.
+        """
+        self.logfile = tempfile.NamedTemporaryFile(
+            prefix="log_sync", dir=logdir, suffix=".log", delete=False)
+
+    def sync_up(self, source, target):
+        return self.execute(self.sync_up_template, source, target)
+
+    def sync_down(self, source, target):
+        return self.execute(self.sync_down_template, source, target)
+
+    def execute(self, sync_template, source, target):
+        """Executes sync_template on source and target."""
+        if self.sync_process:
+            self.sync_process.poll()
+            if self.sync_process.returncode is None:
+                logger.warning("Last sync is still in progress, skipping.")
+                return False
+        final_cmd = sync_template.format(
+            source=quote(source), target=quote(target))
+        logger.debug("Running sync: {}".format(final_cmd))
+        self.sync_process = subprocess.Popen(
+            final_cmd, shell=True, stderr=subprocess.PIPE, stdout=self.logfile)
+        return True
+
+    def wait(self):
+        if self.sync_process:
+            _, error_msg = self.sync_process.communicate()
+            error_msg = error_msg.decode("ascii")
+            code = self.sync_process.returncode
+            self.sync_process = None
+            if code != 0:
+                raise TuneError("Sync error ({}): {}".format(code, error_msg))
+
+    def reset(self):
+        if self.sync_process:
+            logger.warning("Sync process still running but resetting anyways.")
+            self.sync_process = None
+
+    @staticmethod
+    def _validate_sync_string(sync_string):
+        if not isinstance(sync_string, str):
+            raise TuneError("{} is not a string.".format(sync_string))
+        if "{source}" not in sync_string:
+            raise TuneError("Sync template missing '{source}'.")
+        if "{target}" not in sync_string:
```

Comment:
why change this?

Reply:
these were originally ValueErrors

Reply:
whoops was trying to get tests to pass and forgot to revert